File: src/lib/stickerize/fetch-remote.ts
import { lookup } from "node:dns/promises";
import { isIP } from "node:net";
import { badGateway, badRequest, payloadTooLarge, unsupportedMediaType } from "@/lib/stickerize/errors";

const MAX_REDIRECTS = 3;
const ALLOWED_REMOTE_CONTENT_TYPES = new Set(["image/png", "image/jpeg", "image/webp"]);

function isPrivateIpv4(ip: string) {
  const parts = ip.split(".").map((segment) => Number.parseInt(segment, 10));
  if (parts.length !== 4 || parts.some((part) => Number.isNaN(part))) {
    return false;
  }

  const [a, b] = parts;
  return a === 10 || a === 127 || (a === 169 && b === 254) || (a === 172 && b >= 16 && b <= 31) || (a === 192 && b === 168);
}

function isPrivateIpv6(ip: string) {
  const normalized = ip.toLowerCase();
  return normalized === "::1" || normalized.startsWith("fc") || normalized.startsWith("fd") || normalized.startsWith("fe80:") || normalized.startsWith("::ffff:127.") || normalized.startsWith("::ffff:10.") || normalized.startsWith("::ffff:192.168.") || /^::ffff:172\.(1[6-9]|2\d|3[0-1])\./.test(normalized);
}

async function assertSafeRemoteHost(hostname: string) {
  const lower = hostname.toLowerCase();
  if (lower === "localhost" || lower.endsWith(".localhost")) {
    throw badRequest("localhost URLs are not allowed.");
  }

  if (isIP(hostname)) {
    if (isPrivateIpv4(hostname) || isPrivateIpv6(hostname)) {
      throw badRequest("Private network IPs are not allowed.");
    }
    return;
  }

  const records = await lookup(hostname, { all: true, verbatim: true });
  for (const record of records) {
    if ((record.family === 4 && isPrivateIpv4(record.address)) || (record.family === 6 && isPrivateIpv6(record.address))) {
      throw badRequest("Resolved private network IPs are not allowed.");
    }
  }
}

async function readLimitedBody(response: Response, maxBytes: number) {
  const body = response.body;
  if (!body) {
    throw badGateway("Remote image response was empty.");
  }

  const reader = body.getReader();
  const chunks: Uint8Array[] = [];
  let total = 0;

  while (true) {
    const { value, done } = await reader.read();
    if (done) {
      break;
    }

    if (!value) {
      continue;
    }

    total += value.byteLength;
    if (total > maxBytes) {
      throw payloadTooLarge(`Remote images must be smaller than ${maxBytes} bytes.`);
    }

    chunks.push(value);
  }

  return Buffer.concat(chunks.map((chunk) => Buffer.from(chunk)));
}

export async function fetchRemoteImage(imageUrl: string, options: { timeoutMs: number; maxBytes: number; }) {
  let currentUrl = new URL(imageUrl);

  if (currentUrl.protocol !== "https:") {
    throw badRequest("Only https imageUrl values are allowed.");
  }

  for (let redirectCount = 0; redirectCount <= MAX_REDIRECTS; redirectCount += 1) {
    await assertSafeRemoteHost(currentUrl.hostname);

    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), options.timeoutMs);

    try {
      const response = await fetch(currentUrl, {
        redirect: "manual",
        signal: controller.signal,
        headers: {
          Accept: "image/png,image/jpeg,image/webp",
          "User-Agent": "Stickify/1.0",
        },
      });

      if (response.status >= 300 && response.status < 400) {
        const location = response.headers.get("location");
        if (!location) {
          throw badGateway("Remote image redirect did not include a location header.");
        }

        if (redirectCount === MAX_REDIRECTS) {
          throw badGateway("Remote image redirected too many times.");
        }

        currentUrl = new URL(location, currentUrl);
        if (currentUrl.protocol !== "https:") {
          throw badRequest("Redirected image URLs must remain on https.");
        }
        continue;
      }

      if (!response.ok) {
        throw badGateway(`Remote image server responded with ${response.status}.`);
      }

      const contentTypeHeader = response.headers.get("content-type") ?? "";
      const mimeType = contentTypeHeader.split(";")[0].trim().toLowerCase();
      if (!ALLOWED_REMOTE_CONTENT_TYPES.has(mimeType)) {
        throw unsupportedMediaType("Remote URLs must point to a PNG, JPEG, or WebP image.");
      }

      const contentLength = response.headers.get("content-length");
      if (contentLength && Number.parseInt(contentLength, 10) > options.maxBytes) {
        throw payloadTooLarge(`Remote images must be smaller than ${options.maxBytes} bytes.`);
      }

      const buffer = await readLimitedBody(response, options.maxBytes);
      return {
        buffer,
        contentType: mimeType,
        url: currentUrl.toString(),
      };
    } catch (error) {
      if (error instanceof Error && error.name === "AbortError") {
        throw badGateway("Timed out while fetching the remote image.");
      }
      throw error;
    } finally {
      clearTimeout(timeout);
    }
  }

  throw badGateway("Failed to fetch the remote image.");
}

File: backend/app/services/fetch_remote.py
import asyncio
import ipaddress
import socket
from dataclasses import dataclass
from typing import Final
from urllib.parse import urljoin, urlparse

import httpx

from backend.app.core.config import Settings
from backend.app.core.errors import bad_gateway, payload_too_large, unsupported_media_type

ALLOWED_REMOTE_MIME_TYPES: Final[set[str]] = {"image/png", "image/jpeg", "image/webp"}
BLOCKED_HOSTNAMES: Final[set[str]] = {"localhost", "127.0.0.1", "0.0.0.0", "::1"}
MAX_REDIRECTS: Final[int] = 3


@dataclass(slots=True)
class FetchedImage:
    buffer: bytes
    content_type: str


def _is_forbidden_ip(address: str) -> bool:
    ip = ipaddress.ip_address(address.split("%", 1)[0])
    return any(
        (
            ip.is_private,
            ip.is_loopback,
            ip.is_link_local,
            ip.is_multicast,
            ip.is_reserved,
            ip.is_unspecified,
        )
    )


async def _validate_hostname(hostname: str, allow_private_remote_hosts: bool) -> None:
    normalized = hostname.strip().lower().rstrip(".")
    if not normalized:
        raise bad_gateway("The requested host is not allowed.")
    if normalized in BLOCKED_HOSTNAMES and not allow_private_remote_hosts:
        raise bad_gateway("The requested host is not allowed.")

    try:
        loop = asyncio.get_running_loop()
        records = await loop.getaddrinfo(normalized, None, type=socket.SOCK_STREAM)
    except socket.gaierror as error:
        raise bad_gateway("The remote host could not be resolved.") from error

    addresses = {record[4][0] for record in records}
    if not addresses:
        raise bad_gateway("The remote host did not resolve to a routable address.")

    if not allow_private_remote_hosts and any(_is_forbidden_ip(address) for address in addresses):
        raise bad_gateway("The remote host resolved to a blocked network address.")


def _validate_https_url(url: str) -> str:
    parsed = urlparse(url)
    if parsed.scheme != "https":
        raise bad_gateway("Only public https image URLs are supported.")
    if not parsed.hostname:
        raise bad_gateway("The remote image URL is invalid.")
    return url


async def fetch_remote_image(url: str, settings: Settings) -> FetchedImage:
    current_url = _validate_https_url(url)

    async with httpx.AsyncClient(follow_redirects=False, timeout=settings.fetch_timeout_seconds) as client:
        for _ in range(MAX_REDIRECTS + 1):
            parsed = urlparse(current_url)
            await _validate_hostname(
                parsed.hostname or "",
                settings.allow_private_remote_hosts,
            )

            try:
                response = await client.get(
                    current_url,
                    headers={"Accept": "image/png,image/jpeg,image/webp,*/*"},
                )
            except httpx.HTTPError as error:
                raise bad_gateway("Fetching the remote image failed.") from error

            if 300 <= response.status_code < 400:
                location = response.headers.get("location")
                if not location:
                    raise bad_gateway("The remote image redirect was missing a location header.")
                current_url = _validate_https_url(urljoin(current_url, location))
                continue

            if response.status_code >= 400:
                raise bad_gateway(f"The remote image returned HTTP {response.status_code}.")

            content_type = response.headers.get("content-type", "").split(";", 1)[0].strip().lower()
            if content_type not in ALLOWED_REMOTE_MIME_TYPES:
                raise unsupported_media_type("The remote URL did not return a supported image type.")

            content_length = response.headers.get("content-length")
            if content_length and int(content_length) > settings.max_image_bytes:
                raise payload_too_large("The remote image exceeded the configured size limit.")

            payload = bytearray()
            async for chunk in response.aiter_bytes():
                payload.extend(chunk)
                if len(payload) > settings.max_image_bytes:
                    raise payload_too_large("The remote image exceeded the configured size limit.")

            return FetchedImage(buffer=bytes(payload), content_type=content_type)

    raise bad_gateway("The remote image redirected too many times.")
